### ISSUE TITLE
Delay-bind script need an example as it's usually misunderstood

### ISSUE THREAD

**Issue by ili101** [MicrosoftDocs/PowerShell-Docs#9479]
### Prerequisites

- [X] **Existing Issue:** Search the existing issues for this repository. If there is an issue that fits your needs do not file a new one. Subscribe, react, or comment on that issue instead.
- [X] **Descriptive Title:** Write the title for this issue as a short synopsis. If possible, provide context. For example, "Typo in `Get-Foo` cmdlet" instead of "Typo."
- [X] **Verify Version:** If there is a mismatch between documentation and the behavior on your system, ensure that the version you are using is the same as the documentation. Check this box if they match or the issue you are reporting is not version specific.

### Links

https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_script_blocks?view=powershell-7.3#using-delay-bind-script-blocks-with-parameters

### Summary

All examples I found online regarding "delay-bind script" are actually not using it but instead showing the regular automatic variable `$_` 

For example:
https://stackoverflow.com/questions/53326531/delay-bind-script-block-does-not-work-when-function-is-exported-from-module
https://stackoverflow.com/questions/52975186/how-to-pass-psitem-in-a-scriptblock

### Suggested Fix

Add some clarification with examples of what is delay-bind script (and what is not), like:
```powershell
function Test-IsDelayedBinding {
    param(
        [Parameter(ValueFromPipeline)]
        [string]$InputObject
    )

    Process {
        $InputObject
    }
}
# Works
1, 2, 3 | Test-IsDelayedBinding -InputObject { $_ * 2 } 
# Error "A positional parameter cannot be found that accepts argument ' $_ * 2 '."
1, 2, 3 | Test-IsDelayedBinding { $_ * 2 }
# Error "Cannot evaluate parameter 'InputObject' because its argument is specified as a script block and there is no input. A script block cannot be evaluated without input."
Test-IsDelayedBinding { $_ * 2 }

function Test-NotDelayedBinding {
    param(
        [scriptblock]$Process,
        [Parameter(ValueFromPipeline)]
        [string]$InputObject
        
    )

    Process {
        & $Process # $InputObject[n] is passed implicitly as $_ from the "Process" parent scope.
    }
}
# Works
1, 2, 3 | Test-NotDelayedBinding { $_ * 2 } 
# No error
Test-NotDelayedBinding { $_ * 2 }


# Is delayed binding
Get-ChildItem config.log | Rename-Item -NewName { "old_" + $_.Name }
# Is not delayed binding
1, 2, 3 | ForEach-Object { $_ * 2 }
```

**Comment by michaeltlombardi** [comment]
I had a long think about this today and tried to come up with some useful demonstrative example I could give in this section, but what I kept circling back to is that this is a way to _use_ scriptblocks for functions and cmdlets defined by other people (or PowerShell itself).

If you're the cmdlet author and you find yourself needing to handle a particular use case, you're more likely to find it valuable to update your function to handle that use case in detail.

Where delayed binding is really useful is when you _can't_ modify the command you want to use and when you want a more convenient syntax than calling `ForEach-Object` to munge the input when calling the next command in the pipeline.

For instance, the current example in the docs is written as:

```powershell
dir config.log | Rename-Item -NewName {"old_" + $_.Name}
```

Which is more convenient than:

```powershell
dir config.log | ForEach-Object -Process {
  $_ | Rename-Item -NewName "old_$($_.Name)"
}
```

The documentation already defines the circumstances where you can use delayed binding for a cmdlet:

- You must explicitly specify any parameter names you use with delay-bind script blocks.
- The parameter must not be untyped, and the parameter's type cannot be `[scriptblock]` or `[object]`.
- You receive an error if you use a delay-bind script block without providing pipeline input.

So looking at your examples:

- ```powershell
  1, 2, 3 | Test-IsDelayedBinding -InputObject { $_ * 2 } 
  ```
  
  This works because `Test-IsDelayedBinding` defines **InputObject** as a **string**, takes input from the pipeline, and explicitly specifies `-InputObject` in the call.
- ```powershell
  1, 2, 3 | Test-IsDelayedBinding { $_ * 2 }
  ```
  
  This fails because **InputObject** wasn't specified in the call to the function, and the values from the pipeline are already used for the value of **InputObject**.

I'm not seeing a way to expand on this information that will make things clearer. I'm going to close this for now, but if you have an idea for how to add an example or context to clarify this, I'm certainly open to it!

**Comment by ili101** [comment]
I think you are correct that a function example is not needed here, but the fact that I asked for one shows that even when opening the issue I still didn't fully understand delay-bind script.

I didn't understand the documentation after reading it multiple times but reading it now after I know what is dose it looks ok. I assume it's because it's mainly a technical explanation. What is needed is a general simple description before the technical explanation that include what is its purpose, and what is the target audience (function user or author).

For example it may be understood that delay-bind script is something the author of `Rename-Item` implemented and not something you as a function consumer can use on any (suitable) function.

So basically add something like the response you wrote above, for example:
Delay-bind script is a shorthand allowing you to manipulate the pipelined parameters before passing them to a function by using ScriptBlocks.
```powershell
# Renames config.log to old_config.log

# Without delay-bind script 
dir config.log | ForEach-Object {
  $_ | Rename-Item -NewName "old_$($_.Name)"
}
# With delay-bind script 
dir config.log | Rename-Item -NewName {"old_" + $_.Name}
```
